File: R/först_model.R
#============================================================================!
# 
#----------------------------------------------------------------------------!

# start from clean slate
#----------------------------------------------------------------------------!
rm (list = ls ())

# Load necessary dependencies
#----------------------------------------------------------------------------!
source ('R/readData.R')
library ('PopED')

# Get the initial observations
#----------------------------------------------------------------------------!
dataSurv <- readDataSurv ()
dataTime <- readDataTime ()
dataDead <- readDataDead ()

# Boolean: T = comparison to empirical data, F = long term simulation
validationNOTlongterm <- T

# Initialise variables
water_supplies   <- Inf     #
harv_parameters <- c (0, 1) # 
nspecies <- 2               # Beech is species = 1, spruce is species = 2

start_age   <- 0      # age of the stand at the first observation (does not 
                      # affect simulation); not known for mixed stand data
  
nspecies    <- 2                      # number of species in simulation
dx          <- 0.03                   # resolution of stem size dimension [m]
sizes       <- seq (dx, 1.5, by = dx) # stem size scale, in metres
sizes2      <- sizes^2                # square of stem sizes
sizes3      <- sizes^3                # cube of stem sizes
sizeclasses <- length (sizes)         # number of different stem sizes considered
dt          <- 0.01                   # time steps per year

# Determine the length of the simulation
if (validationNOTlongterm) {
  checkpoints <- dataTime - dataTime [1]      # points in simulation time at which empirical data is available
  checkpoints <- checkpoints [-4]             # ignore this one to allow 3x3 plot
  years <- checkpoints [length (checkpoints)] # number of years simulated
} else {
  checkpoints <- c (dataTime [length (dataTime)] - dataTime [1], 125, 200, 300, 500, 1000) #last point for which empirical data is available, and several years after that
  years <- checkpoints [length (checkpoints)]  # number of years simulated
}
timesteps   <- ceiling (years / dt)   # number of time steps in simulation
timesteps   <- 7 / dt

# Set initial values to zero
#----------------------------------------------------------------------------!
Light          <- rep (0, sizeclasses)
TotalLeafArea  <- rep (0, nspecies)
GrowthFlux     <- rep (0, sizeclasses)
GrowthGradient <- rep (0, sizeclasses)
DiameterGrowth <- rep (0, sizeclasses)
DeathRate      <- rep (0, sizeclasses)
u              <- matrix (rep (0, nspecies * sizeclasses), nrow = sizeclasses)
u1             <- rep (0, sizeclasses)
v              <- matrix (rep (0, nspecies * sizeclasses), nrow = sizeclasses)

# Species parameters for 
#     - beech  (first  entry)
#     - spruce (second entry)
#----------------------------------------------------------------------------!
BeechParameters  <- c (0.0055, 2.7, -0.042, 0.000008, 5000.0)
SpruceParameters <- c (0.0072, 1.7, -0.040, 0.000050, 5000.0)
GrowthRate       <- c (BeechParameters [1], SpruceParameters [1]) # species-specific growth rate []
maxsize          <- c (BeechParameters [2], SpruceParameters [2]) # max tree size from www.monumentaltrees.com
RespirationRate  <- GrowthRate / maxsize                          # respiration rate []
WaterCapacity    <- water_supplies                                #
lambda           <- c (BeechParameters [3], SpruceParameters [3]) # proportional to beer-lambert extinction coefficient
DeathPar         <- c (BeechParameters [4], SpruceParameters [4]) #
ReproductionRate <- c (BeechParameters [5], SpruceParameters [5]) # estimated via Yoda's rule

# harvest parameters
#----------------------------------------------------------------------------!
HarvestIndex <- ceiling (harv_parameters [1] * sizeclasses)
RegenYears   <- harv_parameters [2]
HarvestRate  <- 0.0

# set initial size distribution (u) according to first observation
#----------------------------------------------------------------------------!
for (species in 1:nspecies) {
  # empirical diameter data is in mm, convert to meters
  u0 <- 0.001 * c (dataSurv [[species]] [[1]] [, 1], dataDead [[species]] [[1]] [, 1])
  u [1, species] <- sum (u0 <= sizes [1], na.rm = T)
  for (i in 2:sizeclasses) {
    k <- (u0 > sizes [i - 1]) & (u0 <= sizes [i])
    u [i, species] <- round (sum (k, na.rm = T) / dx) # convert number of individuals with diameter between sizes (i-1) and sizes (i) to density
  }
}

# function for ploting size distribution
#----------------------------------------------------------------------------!
pl.sizeDistribution <- function (u, v, sizes, y_max = 4000, title) {
  plot (x     = sizes,
        y     = u [, 1],
        xlab  = 'dbh [m]',
        ylab  = 'number of trees',
        xlim  = c (0, 1.5),
        ylim  = c (0, y_max),
        typ   = 'l',
        col   = 'blue',
        lty   = 2,
        las   = 1,
        main = title)
  lines (x    = sizes,
         y    = u [, 2],
         col  = 'red',
         lty  = 2)
  lines (x    = sizes,
         y    = v [, 1],
         col  = 'blue')
  lines (x    = sizes,
         y    = v [, 2],
         col  = 'red')
}
y_maxs <- c (4000, 2000, 1000, 600, 600, 600, 400, 400, 300)

# Plot initial size distribution
layout (matrix (1:9, nrow = 3, byrow = T))
par (mar = c (5, 5, 3, 1))
pl.sizeDistribution (u = u, v = u, sizes = sizes, y_max = 4000, title = '0 years')

# Set timer to measure execution time
#----------------------------------------------------------------------------!
start.time <- Sys.time ()

# Function to calculate incident light upon the trees of sizeclass i
#calcLayer <- function (species, i, sizes2 = sizes2) {
#  Ind_Light [i] <- Ind_Light [i + 1] + u [i + 1, species] * sizes2 [i + 1]
#}

# Loop over timesteps
#----------------------------------------------------------------------------!
for (t in 1:timesteps) {
  # calculate portion of light (Light) arriving at heights corresponding
  # to different stem diameter sizes
  #--------------------------------------------------------------------------!
  Light <- rep (1, sizeclasses) # Initially assume all are fully illuminated
  for (species in 1:nspecies) {
    Ind_Light <- rep (0, sizeclasses)
    for (i in (sizeclasses-1):1) {
      Ind_Light [i] <- Ind_Light [i + 1] + u [i + 1, species] * sizes2 [i + 1]
    }
    TotalLeafArea [species] <- dx * (Ind_Light [1] + u [1] * sizes2 [1])
    Ind_Light <- Ind_Light + 0.5 * u [, species] * sizes2
    Light <- Light * exp (lambda [species] * dx * Ind_Light)
  }

  # gross photosynthetic production assuming no water constraints, if this 
  # exceeds WaterCapacity, then all actual growth is reduced accordingly
  PotentialProduction <- 0
  for (species in 1:nspecies) {
    PotentialProduction <- PotentialProduction + sum (GrowthRate [species] * Light * u [i, species] * sizes2) * dx  
  }

  for (species in 1:nspecies) {
    # calculate the different components determining the size distribution
    # one time step later. this is done as described in the paper

    # Growth
    DiameterGrowth = (GrowthRate [species] * Light * min (1, WaterCapacity / 
                                                             PotentialProduction) 
                     - RespirationRate [species] * sizes)

    # Death
    MassGrowth <- DiameterGrowth * sizes2
    DeathRate  <- 0.5 * DeathPar [species] / pmax (0, MassGrowth)
    DeathRate [is.nan (DeathRate)] <- 0

    end <- sizeclasses
    # The error seemed to have been downstream of this
    GrowthFlux                 <- DiameterGrowth * u [, species]
    GrowthGradient [1]         <- (GrowthFlux [2]     - GrowthFlux [1])         / dx
    GrowthGradient [end]       <- (GrowthFlux [end]   - GrowthFlux [end-1])     / dx
    GrowthGradient [2:(end-1)] <- (GrowthFlux [3:end] - GrowthFlux [1:(end-2)]) / (2 * dx);

    #PDE
    #"max(0," isnt really necessary here, but an efficient way to force stability
    u1     <- max (0, u[ , species] - dt * (GrowthGradient + (DeathRate + HarvestRate) * u [, species])) 
    u1 [1] <- ReproductionRate [species] * TotalLeafArea [species] # birth of new trees
    u [, species] <- u1
  }

  # Plot the size distribution
  k0 <- which (t * dt == checkpoints) # checkpoints [k0] = t*dt, if there is such a k0
  if (!isempty (k0)) {
    if (validationNOTlongterm) {
      for (species in 1:nspecies) {
        # empirical diameter data is in mm, convert to meters
        v0 <- 0.001 * c (dataSurv [[species]] [[1]] [, k0], dataDead [[species]] [[1]] [, k0])
        v [1, species] <- round (sum (v0 <= sizes [1], na.rm = T) / dx)
        for (i in 2:sizeclasses) {
          k <- (v0 > sizes [i - 1]) & (v0 <= sizes [i])
          v [i, species] <- round (sum (k, na.rm = T) / dx) # convert number of individuals with diameter between sizes (i-1) and sizes (i) to density
        }
      }
      print (c (u, v))
      pl.sizeDistribution (u = u, v = v, sizes = sizes, y_max = y_maxs [k0],
                           title = paste (as.character (checkpoints [k0]), ' years', sep = ''))     
    }
  } # Still need to introduce plotting functionality for long-term simulations
}

# Determine time it took to run
#----------------------------------------------------------------------------!
end.time   <- Sys.time ()
time.taken <- end.time - start.time
time.taken
#============================================================================!

File: fortran/först_model.f90
!============================================================================!
subroutine forst_model (u0,       & ! initial size distribution
                        ut,       & ! final size distribution
                        years,    & ! number of years in simulation
                        !dt,       & ! time steps per year
                        nspecies  & ! number of species in simulation

                        )
!============================================================================!
! The först model simulates forest size distribution employing a set of 
! partial differential equations (PDE)
!----------------------------------------------------------------------------!
!     developed by:
!     Robert Beyer    (rb792@cam.ac.uk)
!     Tim Rademacher  (trademacher@fas.harvard.edu)
!    
!     initially parameterised and evaluated against permanent plot 
!     measurements of beech and spruce from the Technische Universität 
!     München (TUM) from 1848? to present.
!----------------------------------------------------------------------------!

! start from clean slate
!----------------------------------------------------------------------------!
implicit none

! declare and initialise variables
!----------------------------------------------------------------------------!
integer :: i                                  ! loop variable
integer :: species                            ! loop variable
integer, intent (in) :: nspecies              ! number of species in simulation
real (kind = 8), parameter :: dx      = 0.03  ! resolution of stem size dimension [m]
integer, parameter :: nsizes = int (dx / 1.5) ! number of size classes up to 1.5m     
real (kind = 8), dimension (nsizes) :: sizes  = (/ (real (i*dx), i = 1, nsizes) /) ! Different size classes
real (kind = 8), dimension (nsizes) :: sizes2 = (/ (sizes (i)**2.0, i = 1, nsizes) /) ! Size classes squared
real (kind = 8), dimension (nsizes) :: sizes3 = (/ (sizes (i)**3.0, i = 1, nsizes) /) ! Size classes cubed
real (kind = 8), dimension (nspecies, nsizes) :: u0, u1, u2, ut ! initial, transient and final size distributions
!----------------------------------------------------------------------------!
integer, intent (in) :: years                ! number of years in simulation
real (kind = 8), parameter :: dt = 0.01      ! time step as proportion of year
integer :: t                                 ! loop variable for the time step 
integer :: timesteps                         ! number of time steps in simulation

real (kind = 8), dimension (nsizes) :: light ! proportion of light penetrating to each size class
real (kind = 8), dimension (nsizes) :: indLight !

real (kind = 8), dimension (5), parameter :: beechParameters  = (/0.0055, 2.7, -0.042, 0.000008, 5000.0/)
real (kind = 8), dimension (5), parameter :: spruceParameters = (/0.0072, 1.7, -0.040, 0.000050, 5000.0/)
real (kind = 8), dimension (nspecies) :: totalLeafArea      !
real (kind = 8), dimension (nspecies) :: lambda     		!
real (kind = 8), dimension (nspecies) :: growthRate         ! 
real (kind = 8), dimension (nspecies) :: maxSize            ! maximum tree stem diameter [m]
real (kind = 8), dimension (nspecies) :: respirationRate    !
real (kind = 8), dimension (nspecies) :: deathPar           !
real (kind = 8), dimension (nspecies) :: reproductionRate   !

real (kind = 8), parameter :: waterSupplies = huge (0.0d0) ! Set water supply to infinity
real (kind = 8), dimension (nsizes) :: waterCapacity   = waterSupplies
real (kind = 8), dimension (nsizes) :: growthFlux      = 0 !
real (kind = 8), dimension (nsizes) :: growthGradient  = 0 !
real (kind = 8), dimension (nsizes) :: diameterGrowth  = 0 !
real (kind = 8), dimension (nsizes) :: deathRate       = 0 !
real (kind = 8), dimension (nsizes) :: massGrowth      = 0 !
real (kind = 8) :: potentialProduction                !

! harvest parameters
!----------------------------------------------------------------------------!
real (kind = 8), dimension (2), parameter :: harvParameters = (/0, 1/)
real (kind = 8), parameter :: harvestIndex = ceiling (harvParameters (1) * nsizes) ! nsizes or size classes
real (kind = 8), parameter :: regenYears   = harvParameters (2)
real (kind = 8), parameter :: harvestRate = 0.0

! initialise variables
!----------------------------------------------------------------------------!
totalLeafArea = 0

! set species-specific parameters
!----------------------------------------------------------------------------!
if (nspecies == 1) then
  growthRate       = beechParameters (1)
  maxSize          = beechParameters (2)
  lambda           = beechParameters (3)
  deathPar         = beechParameters (4)
  reproductionRate = beechParameters (5)
else if (nspecies == 2) then
  growthRate       = (/beechParameters (1), spruceParameters (1)/)
  maxSize          = (/beechParameters (2), spruceParameters (2)/)
  lambda           = (/beechParameters (3), spruceParameters (3)/)
  deathPar         = (/beechParameters (4), spruceParameters (4)/)
  reproductionRate = (/beechParameters (5), spruceParameters (5)/)
end if
respirationRate = growthRate / maxSize

! calculate the number of time steps
!----------------------------------------------------------------------------!
timesteps = ceiling (years / dt)

! set initial size distribution
!----------------------------------------------------------------------------!
u1 = u0

! loop over timesteps
!----------------------------------------------------------------------------!
time : do t = 1, timesteps
  ! calculate portion of light (Light) arriving at heights corresponding
  ! to different stem diameter sizes
  !--------------------------------------------------------------------------!
  light = 1    ! Initially assume all are fully illuminated
  calcLight : do species = 1, nspecies
    indLight = 0 ! Set light penetration to 0  
    layers : do  i = nsizes-1, 1, -1 ! Loop over canopy layers
      indLight (i) = indLight (i + 1) + u1 (i + 1, species) * sizes2 (i + 1)
    end do layers ! loop ending
    totalLeafArea (species) = dx * (indLight (1) + u1 (1, species) * sizes2 (1)) ! Not sure u is correct here 
    indLight = indLight + 0.5 * u1 (:, species) * sizes2
    light    = light * exp (lambda (species) * dx * indLight)
  end do calcLight ! loop ending

  ! gross photosynthetic production assuming no water constraints, if this 
  ! exceeds WaterCapacity, then all actual growth is reduced accordingly
  !--------------------------------------------------------------------------!
  potentialProduction = 0
  calcPotentialProduction : do species = 1, nspecies
    potentialProduction = potentialProduction +                              &
                          sum (growthRate (species) * light *                &
                               u1 (:, species) * sizes2) * dx  
  end do calcPotentialProduction ! loop ending

  calcSizeDistribution : do species = 1, nspecies
    ! calculate the different components determining the size distribution
    ! one time step later. this is done as described in the paper

    ! Growth
    !------------------------------------------------------------------------!
    diameterGrowth = (growthRate (species) * light *                         &
                      min (1.0, waterCapacity / potentialProduction) -       &
                      respirationRate (species) * sizes)

    ! Death
    massGrowth = diameterGrowth * sizes2
    deathRate  = 0.5 * deathPar (species) / max (0.0, massGrowth)
    where (isnan (deathRate)) deathRate = 0

    growthFlux                    = diameterGrowth * u1 (:, species)
    growthGradient (1)            = (growthFlux (2)        - growthFlux (1))         / dx
    growthGradient (nsizes)       = (growthFlux (nsizes)   - growthFlux (nsizes-1))     / dx
    growthGradient (2:(nsizes-1)) = (growthFlux (3:nsizes) - growthFlux (1:(nsizes-2))) / (2 * dx);

    !Partial Differential Equation
    !"max(0," isnt really necessary here, but an efficient way to force stability
    u2              = max (0, u (:, species) - dt * (GrowthGradient + (DeathRate + HarvestRate) * u1 (:, species))) 
    u2 (1, species) = ReproductionRate (species) * TotalLeafArea (species) ! birth of new trees
    u1              = u2
  end do calcSizeDistribution ! loop ending

  ut = u1
  write (*, *) 'Done with loop ',t
end do time ! loop ending

!============================================================================!
end subroutine
!============================================================================!